[FEATURE] EPIX Réinitialisation en masse des mots de passe temporaires des élèves sur Pix Orga (PIX-8678)

igorissen opened the pull request:
## :unicorn: Problème

En début d'année scolaire, des élèves peuvent avoir oublié leurs identifiants et/ou mots de passe. Les professeurs doivent exécuter les différentes étapes suivantes dans Pix Orga pour fournir les identifiants ainsi que les nouveaux mots de passe à chacun des élèves concernés :

- Retrouver les élèves
- Ouvrir la fenêtre de gestion du compte d'un élève
- Cliquer sur l'option de réinitialisation de mot de passe
- Fournir les informations identifiant et nouveau mot de passe à l'élève avant de ferme la fenêtre

## :robot: Proposition

Permettre aux professeurs de réinitialiser le mot de passe de plusieurs élèves en même temps au lieu de le faire un par un. Les professeurs pourront récupérer un fichier au format CSV une fois que l'action de réinitialisation en masse s'est bien déroulée. Ce fichier contiendra les informations suivantes :

- Classe
- Nom
- Prénom
- Identifiant
- Mot de passe (à usage unique)

## :rainbow: Remarques

- Les organisations ayant le type `SCO` et la propriété `isManagingStudents` à `true` peuvent avoir accès à cette option de réinitialisation en masse.
- Seul les élèves ayant une méthode d'authentification `identifiant` (connexion possible avec un `username`) peuvent avoir leur mot de passe réinitialisé
- ⚠️  Les professeurs ne pourront réinitialiser que les élèves qui sont sélectionnés sur une seule page. La gestion de plusieurs page dû au grand nombre d'élèves n'est pas pris en compte.

## :100: Pour tester

Les RA : 
  * [Orga (.fr)](https://orga-pr6566.review.pix.fr)
  * [Orga (.org)](https://orga-pr6566.review.pix.org)

Suivre les instructions de chacune des PR suivantes pour tester la fonctionnalité dans son ensemble :

1. #6561
2. #6567
3. #6568
4. #6574
5. #6603
6. #6682

igorissen committed:
feat(orga): add TableHeaders component

will be used inside ScoOrganizationParticipant::List component

Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

igorissen committed:
feat(orga): add TableRow component

will be used inside ScoOrganizationParticipant::List component

Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

igorissen committed:
feat(orga): display multi-select checkboxes

Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

er-lim committed:
feat(orga): Add action bar component in sco organization participant page

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>

igorissen committed:
feat(orga): display reset password modal

Co-authored-by: Eric Lim <eric.lim@zenika.com>

igorissen committed:
feat(api): add batchUpdatePasswordThatShouldBeChanged repository method

Co-authored-by: Eric Lim <eric.lim@zenika.com>

er-lim committed:
feat(api): add getByIds method in user repository

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>

igorissen committed:
feat(api): add update organization learners password usecase

Co-authored-by: Eric Lim <eric.lim@zenika.com>

igorissen committed:
feat(api): add update organization learners password method in sco-organization-learners controller

Co-authored-by: Eric Lim <eric.lim@zenika.com>

igorissen committed:
feat(orga): reset students password

Co-authored-by: Eric Lim <eric.lim@zenika.com>

igorissen committed:
feat(api): return list of organization learners generated password

Co-authored-by: Eric Lim <eric.lim@zenika.com>

igorissen committed:
feat(api): add writeCsvUtils as dependency to inject

igorissen committed:
refactor(api): rename imported module name

igorissen committed:
feat(api): add generatedResetOrganizationLearnersPasswordCsv usecase

igorissen committed:
feat(api): return generated CSV content

igorissen committed:
feat(api): use a domain transaction

igorissen committed:
refactor(api): fix flaky test

igorissen commented on orga/translations/fr.json:
```diff
@@ -1009,6 +1009,7 @@
           },
           "last-participation-date": "Dernière participation",
           "login-method": "Méthode(s) de connexion",
+          "mainCheckbox": "Selectionner/Deselectionner toute la colonne",
```
```suggestion
          "mainCheckbox": "Sélectionner/Désélectionner toute la colonne",
```

igorissen commented on orga/translations/fr.json:
```diff
@@ -983,6 +983,7 @@
       "page-title": "Élèves",
       "table": {
         "column": {
+          "checkbox": "Sélectionner/Déselectionner {firstname} {lastname}",
```
```suggestion
          "checkbox": "Sélectionner/Désélectionner {firstname} {lastname}",
```

igorissen commented on orga/app/components/sco-organization-participant/action-bar.hbs:
```diff
@@ -0,0 +1,10 @@
+    {{t "pages.sco-organization-participants.action-bar.information" count=@count}}
+  </:information>
+  <:actions>
+    <PixButton @triggerAction={{@openResetPasswordModal}} type="button" @backgroundColor="red">
```
```suggestion
    <PixButton @triggerAction={{@openResetPasswordModal}} type="button">
```

igorissen commented on orga/tests/integration/components/sco-organization-participant/list_test.js:
```diff
@@ -1103,4 +1103,56 @@ module('Integration | Component | ScoOrganizationParticipant::List', function (h
+  module('action bar', function () {
+    test('displays action bar', async function (assert) {
+      //given
+      const students = [{ id: 1, firstName: 'Spider', lastName: 'Man' }];
```
```suggestion
      const students = [{ id: 1, firstName: 'Spider', lastName: 'Man' }, { id: 2, firstName: 'Spider', lastName: 'Woman' }];
```

igorissen committed:
feat(orga): create CSV file from api response

Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>

lego-technix commented on orga/translations/fr.json:
```diff
@@ -890,6 +891,10 @@
       "title": "{count, plural, =0 {Élèves} =1 {Élève ({count})} other {Élèves ({count})}}",
+      "action-bar": {
+        "information": "{count, plural, =1 {{count} élève sélectionné} other {{count} élèves sélectionnés}}",
+        "reset-password-button": "Réinitialiser les mots de passes des élèves sélectionnés"
```
```suggestion
        "reset-password-button": "Réinitialiser les mots de passe des élèves sélectionnés"
```

lego-technix commented on orga/translations/fr.json:
```diff
@@ -983,8 +988,16 @@
       "no-participants-action": "L’administrateur doit importer la base élèves en cliquant sur le bouton importer.",
       "page-title": "Élèves",
+      "reset-password-modal": {
+        "title": "Réinitialisation de mot de passe des élèves avec identifiants",
```
```suggestion
        "title": "Réinitialisation de mot de passe des élèves avec identifiant",
```

lego-technix commented on orga/translations/fr.json:
```diff
@@ -983,8 +988,16 @@
+        "content-message-1": "Sur {totalSelectedStudents, plural, =1 {<strong>{totalSelectedStudents} élève</strong>} other {<strong>{totalSelectedStudents} élèves</strong>}} que vous avez sélectionné, {totalAffectedStudents, plural, =0 {<strong>aucun mot de passe</strong> sera réinitialisé} =1 {le mot de passe de <strong>{totalAffectedStudents} élève</strong> va être réinitialisé.} other {les mo ...
+        "content-message-2": "Un fichier CSV sera généré contenant les identifiants ainsi que les mots de passe à usage unique.",
+        "content-message-3": "Communiquez ces nouveaux mots de passe à usage unique aux élèves. Lorsqu'ils se connecteront avec ce mot de passe, Pix leur demandera d'en saisir un nouveau.",
+        "warning-message": "Seuls les mots de passe des accès par identifiants peuvent être réinitialisés. Les accès par adresse email et Mediacentre ne seront pas impactés."
```
```suggestion
        "warning-message": "Seuls les mots de passe des accès par identifiant peuvent être réinitialisés. Les accès par adresse email et Mediacentre ne sont pas impactés."
```

lego-technix commented on orga/translations/fr.json:
```diff
@@ -983,8 +988,16 @@
       "page-title": "Élèves",
+      "reset-password-modal": {
+        "title": "Réinitialisation de mot de passe des élèves avec identifiants",
+        "content-message-1": "Sur {totalSelectedStudents, plural, =1 {<strong>{totalSelectedStudents} élève</strong>} other {<strong>{totalSelectedStudents} élèves</strong>}} que vous avez sélectionné, {totalAffectedStudents, plural, =0 {<strong>aucun mot de passe</strong> sera réinitialisé} =1 {le mot de passe de <strong>{totalAffectedStudents} élève</strong> va être réinitialisé.} other {les mo ...
```
```suggestion
        "content-message-1": "Sur {totalSelectedStudents, plural, =1 {<strong>{totalSelectedStudents} élève</strong> sélectionné} other {<strong>{totalSelectedStudents} élèves</strong> sélectionnés}}, {totalAffectedStudents, plural, =0 {<strong>aucun mot de passe</strong> ne sera réinitialisé} =1 {le mot de passe de <strong>{totalAffectedStudents} élève</strong> va être réinitialisé} other {les mots de passe de <strong>{totalAffectedStudents} élèves</strong> vont être réinitialisés}}.",
```

lego-technix commented on api/lib/domain/errors.js:
```diff
@@ -987,8 +987,11 @@ class UserNotAuthorizedToCertifyError extends DomainError {
-  constructor(message = "L'utilisateur n'est pas autorisé à mettre à jour ce mot de passe.") {
-    super(message);
+  constructor(
+    message = "L'utilisateur n'est pas autorisé à mettre à jour ce mot de passe.",
```
Ne pas mettre de message en français dans l'API

lego-technix commented on api/lib/application/sco-organization-learners/index.js:
```diff
@@ -204,6 +204,43 @@ const register = async function (server) {
       },
     },
+    {
+      method: 'PUT',
+      path: '/api/sco-organization-learners/passwords',
```
Étant donné qu'il ne s'agit pas de la mise à jour d'une ressource qui serait `passwords` mais plutôt d'une API pour workflow c'est plutôt un verbe qu'il faut utiliser et avec un POST :

```suggestion
      method: 'POST',
      path: '/api/sco-organization-learners/password-reset',
```

L'action est *password-reset* et on lui passe de 1 à n identifiants, _passwords-reset_ au pluriel serait moins adapté.

er-lim replied:
c'est bon pour moi. Je remets l'article que tu m'as partagé pour des infos si ça intéresse d'autres personnes : https://www.thoughtworks.com/insights/blog/rest-api-design-resource-modeling 😄

lego-technix commented on api/lib/application/sco-organization-learners/index.js:
```diff
@@ -204,6 +204,43 @@ const register = async function (server) {
+          },
+        },
+        notes: [
+          "- Met à jour les mots de passe des élèves ayant un identifiant comme méthode d'authentification\n" +
+            "- La demande de modification du mot de passe doit être effectuée par un membre de l'organisation à laquelle appartient les élèves.",
```
```suggestion
          "- Réinitialise, avec un mot de passe à usage unique, les mots de passe des élèves dont les identifiants sont passés en paramètre et qui ont un identifiant comme méthode d'authentification\n" +
            "- La demande de modification du mot de passe doit être effectuée par un membre de l'organisation à laquelle appartiennent les élèves.",
```

lego-technix commented on api/lib/domain/constants/update-organization-learners-password-errors.js:
```diff
@@ -0,0 +1,10 @@
+const ORGANIZATION_LEARNER_WITHOUT_USERNAME_CODE = 'ORGANIZATION_LEARNER_WITHOUT_USERNAME';
+const USER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE = 'USER_DOES_NOT_BELONG_TO_ORGANIZATION';
+const ORGANIZATION_LEARNER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE =
+  'ORGANIZATION_LEARNER_DOES_NOT_BELONG_TO_ORGANIZATION';
+
+export {
+  ORGANIZATION_LEARNER_WITHOUT_USERNAME_CODE,
+  ORGANIZATION_LEARNER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE,
+  USER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE,
+};
```
Mettre un ordre plus logique pour les codes d'erreur en mettant l'ordre dans lequel les erreurs peuvent arriver : 
```suggestion
const USER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE = 'USER_DOES_NOT_BELONG_TO_ORGANIZATION';
const ORGANIZATION_LEARNER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE =
  'ORGANIZATION_LEARNER_DOES_NOT_BELONG_TO_ORGANIZATION';
const ORGANIZATION_LEARNER_WITHOUT_USERNAME_CODE = 'ORGANIZATION_LEARNER_WITHOUT_USERNAME';

export {
  USER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE,
  ORGANIZATION_LEARNER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE,
  ORGANIZATION_LEARNER_WITHOUT_USERNAME_CODE,
};
```

lego-technix commented on api/lib/domain/usecases/update-organization-learners-password.js:
```diff
@@ -0,0 +1,79 @@
+import { UserNotAuthorizedToUpdatePasswordError } from '../errors.js';
+import {
+  ORGANIZATION_LEARNER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE,
+  ORGANIZATION_LEARNER_WITHOUT_USERNAME_CODE,
+  USER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE,
```
Mettre un ordre plus logique pour les codes d'erreur en mettant l'ordre dans lequel les erreurs peuvent arriver : 

```suggestion
  USER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE,
  ORGANIZATION_LEARNER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE,
  ORGANIZATION_LEARNER_WITHOUT_USERNAME_CODE,
```

lego-technix commented on api/lib/application/sco-organization-learners/sco-organization-learner-controller.js:
```diff
@@ -180,6 +182,33 @@ const checkScoAccountRecovery = async function (
+    });
+    return usecases.generateResetOrganizationLearnersPassword({
+      organizationLearnersGeneratedPassword,
+    });
```
Renommer ce usecase pour indiquer clairement qu'il génère un fichier CSV : 

```suggestion
    return usecases.generateResetOrganizationLearnersPasswordCsvContent({
      organizationLearnersGeneratedPassword,
    });
```

lego-technix commented on api/lib/application/sco-organization-learners/sco-organization-learner-controller.js:
```diff
@@ -180,6 +182,33 @@ const checkScoAccountRecovery = async function (
+      organizationId,
+      organizationLearnersId,
+      domainTransaction,
+    });
```
Renommer ce usecase pour indiquer clairement qu'il s'agit de la fonction de _reset password_ qui est le terme métier qui doit être utilisé partout :

```suggestion
    const organizationLearnersGeneratedPassword = await usecases.resetOrganizationLearnersPassword({
      userId,
      organizationId,
      organizationLearnersId,
      domainTransaction,
    });
```

lego-technix commented on api/lib/domain/usecases/update-organization-learners-password.js:
```diff
@@ -0,0 +1,79 @@
+          password: generatedPassword,
+          division: organizationLearnersMap.get(userId).division,
+        })
+      );
```
Utilisation d'une variable plus parlante contenant le terme métier _reset password_ 
:
```suggestion
  const organizationLearnersPasswordResets = [];

  const usersToUpdateWithNewPassword = await Promise.all(
    students.map(async ({ id: userId, username }) => {
      const generatedPassword = passwordGenerator.generateSimplePassword();
      const hashedPassword = await encryptionService.hashPassword(generatedPassword);

      organizationLearnersPasswordResets.push(
        new OrganizationLearnerPasswordResetDTO({
          username,
          password: generatedPassword,
          division: organizationLearnersMap.get(userId).division,
        })
      );
```

lego-technix committed:
feat(orga): fix translations

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: BKsensei <rebecca.kaci@gmail.com>

lego-technix committed:
feat(orga): delete password reset button unnecessary background color

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: BKsensei <rebecca.kaci@gmail.com>

lego-technix committed:
feat(api): update reset password route description

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: BKsensei <rebecca.kaci@gmail.com>

lego-technix committed:
feat(orga): fix linting

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: BKsensei <rebecca.kaci@gmail.com>

lego-technix committed:
feat(api): fix linting

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: BKsensei <rebecca.kaci@gmail.com>

reibecca committed:
test(orga): add second students to test

reibecca committed:
re
factor(api): lint tests section in user-repository-test

reibecca committed:
refactor(orga): lint tests section list-test

reibecca committed:
refactor(api): reorganize error codes

reibecca committed:
refactor(api): reorganize error codes bis

reibecca committed:
refactor(api): rename usecase to be clearer

reibecca committed:
refactor(api): rename usecase to be clearer bis

reibecca committed:
refactor(api): rename usecase to be clearer bis-bis

lego-technix committed:
feat(api): change API route to be REST workflow

lego-technix committed:
feat(orga): change API route to be REST workflow

lego-technix committed:
refactor(orga): add the possibility to use HTML in modal title + change confirmation button background color

Co-authored-by: BKsensei <rebecca.kaci@gmail.com>

lego-technix committed:
feat(orga): display notifications for password reset success or error

Co-authored-by: BKsensei <rebecca.kaci@gmail.com>
Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

reibecca committed:
feat(orga): do not display content messages 2 et 3 in modal when no students with username are selectionned

lego-technix committed:
feat(orga): improve condition wording

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: BKsensei <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

lego-technix committed:
feat(api): add more info in CSV export

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: BKsensei <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

igorissen committed:
refactor(api): remove useless organization membership check

because this is already done by a prehandler (securityPreHandlers.checkUserBelongsToScoOrganizationAndManagesStudents)

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

lego-technix reviewed (approved):
✅ Testé fonctionnellement avec succès avec Firefox

igorissen reviewed (commented):
Tests ✅ sur Safari et Safari Technology Preview

reibecca commented:
✅ Tester sur les RA sur Chrome 🎉
